### PR TITLE
Fix path separator used by normalizeTarEntry() under Windows

### PR DESCRIPTION
**Pull request by pdcastro**
Connects-to: https://github.com/balena-io/balena-cli/issues/1133
Change-type: patch
Signed-off-by: Paulo Castro <paulo@balena.io>

### CODE REVIEW

**Comment by pdcastro** [comment]
I've just force-pushed a complete replacement of the previous code, because I when I actually tested it on Windows, it failed. @zvin and others, apologies for having placed the original PR for review before having tested it on Windows (luckily it was a very small change...). What is now up for review has been tested on Windows and MacOS. Thanks!

**Comment by pdcastro** [comment]
OK, one more change. The tests were passing on Windows and MacOS, yes, but then I thought of additional test cases which were failing on all platforms:
```
input → expected output
'' → ''
'.' → '.'
'..' → '..'
'/' → '.'
'/.' → '.'
'/..' → '.'
'./' → '.'
'../' → '..'
'/./' → '.'
'/../' → '.'
```
(At least some of those were failing.) **Also I have added removal of any trailing slashes.** The purpose of normalization is to provide a single representation for multiple ways of referring to the same entity, so for example directories `'foo/bar/'` and `'foo/bar'` are now both normalized as `'foo/bar'` (the tar entry header has a specific field for the file type, to distinguish between files and folders, so a trailing slash is not required).

I think the trailing slash removal, however, could be seen as a breaking change (?), and for this reason I have set the change-type to 'major'.

File: lib/index.ts [hunk 1]
```diff
@@ -29,13 +43,24 @@ const noop = () => {};
...
 export function normalizeTarEntry(name: string): string {
-	const normalized = path.normalize(name);
-	if (path.isAbsolute(normalized)) {
-		return normalized.substr(normalized.indexOf(path.sep) + 1);
+	if (!name) {
+		return '';
+	}
+	let normalized = normalize(name);
+	// remove any leading or trailing slashes
+	if (normalized.startsWith('/')) {
```
Comment (hedss) [inline lib/index.ts]: I don't know how much other code in here would benefit from `lodash`, but certainly `return _.trim(normalized, '/')` would save the next 7 lines.

File: lib/index.ts [hunk 2]
```diff
@@ -225,3 +250,12 @@ export function multicastStream(
 		);
 	});
 }
+
+/**
+ * Escape characters that have a special meaning in regular expressions.
+ * Borrowed from:
+ * https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions
+ */
+function escapeRE(s: string) {
```
Comment (hedss) [inline lib/index.ts]: I guess `lodash` would be useful then, as there's a `_.escapeRegExp()` method that does exactly this.

File: lib/index.ts [hunk 3]
```diff
@@ -20,6 +20,20 @@ import * as tar from 'tar-stream';
...
+/**
+ * Alternative path.normalize() function that uses '/' as a path separator in
+ * non-POSIX systems like Windows (which uses a backslash). This is useful for
+ * tar files/streams as the path separator should always be '/', regardless of
+ * platform/OS, as per tar spec:
+ * https://www.gnu.org/software/tar/manual/html_node/Standard.html
+ */
+function normalizeNonPosix(p: string): string {
+	return path.normalize(p).replace(pathSepRE, '/');
+}
```
Comment (zvin) [inline lib/index.ts]: You can use `path.posix.normalize` here.

File: lib/index.ts (hunk 1 above)
Comment (pdcastro) [reply lib/index.ts]: Nice! There was also a global `noop` function that I have now replaced with `_.noop`. (In the past I thought that `noop` alone wasn't enough reason to add lodash as a dependency, but `_.trim` tipped the balance for me.) :-)

File: lib/index.ts (hunk 2 above)
Comment (pdcastro) [reply lib/index.ts]: Well, after @zvin's comment, it turns out I don't need a regular expression at all...

File: lib/index.ts (hunk 3 above)
Comment (pdcastro) [reply lib/index.ts]: What?!? Where did that come from?! I had just saved several lines of code using lodash following @hedss comment, now I can delete a couple more functions too. What next? I guess if I just wait another half an hour, someone else will add a comment: _"oh, you're using `tar-utils`? you don't need that, just `import genie` and you're done!"_ or maybe: _"oh, all of that to fix the CLI under Windows? you don't need the CLI, just say 'Siri, push the app to balena' and it's all sorted"._
LOL :-)

Comment (zvin) [reply lib/index.ts]: Unfortunately, we're not using python https://xkcd.com/353/ :)

File: lib/index.ts [hunk 4]
```diff
@@ -55,9 +62,9 @@ export function streamToBuffer(stream: Readable): Bluebird<Buffer> {
 	})
 		.then(Buffer.concat)
 		.finally(() => {
-			stream.removeListener('data', onData || noop);
-			stream.removeListener('error', onError || noop);
-			stream.removeListener('end', onEnd || noop);
+			stream.removeListener('data', onData || _.noop);
+			stream.removeListener('error', onError || _.noop);
+			stream.removeListener('end', onEnd || _.noop);
```
Comment (pdcastro) [inline lib/index.ts]: Now that `lodash` was added as a dependency, might as well use its `_.noop`.

File: tests/tests.ts [hunk 5]
```diff
@@ -124,7 +150,7 @@ describe('multicastStream', function() {
 		const nStreams = 3;
 		for (let i = 0; i < nStreams; ++i) {
 			toStreams.push(
-				new MockWritable({ highWaterMark: hwm / (i + 1) }, i * 10),
+				new MockWritable({ highWaterMark: Math.ceil(hwm / (i + 1)) }, i * 10),
```
Comment (pdcastro) [inline tests/tests.ts]: This change prevents a non-integer `highWaterMark` value like 16384 / 3 = 5461.3. Oddly the test was passing under MacOS and Linux, but failing under Windows.

**Pull request merged**